### ISSUE TITLE
Please do check if group value is not empty

### ISSUE THREAD

**Issue by parol-peter** [rails/rails#19873]
If somehow query method group is being called with nil param this value goes to group_values and then sql built with ActiveRecord::Calculations.execute_grouped_calculation breaks up

check code like this: Model.group('id').group(nil).count

quick fix is

("#{field} AS #{aliaz}" if field and aliaz) in ActiveRecord::Calculations line 312

**Comment by rafaelfranca** [comment]
I'd say we just just raise an `ArgumentError` if `nil` is passed to `group`. Could you open a PR?

**Comment by parol-peter** [comment]
I believe gracefull ignore for that case would be a way better idea. ofter group value is set dynamically within chain of query methods. Such an error would just mess up an AR querying code and avoid query methods chaining :/ sounds bad. same thing for another query methods...

**Comment by rafaelfranca** [comment]
I think it is a really bad idea ignore bad input. I don't see any reason why you would want to pass `nil` to a `group` method. Could you explain?

**Comment by meinac** [comment]
Currently if you use only one group clause with nil parameter it ignores that bad input.
The error here is using multiple group clause with at least one bad input. If the behaviour of group method with nil parameter is ignoring that input so IMHO it should ignore that bad input with multiple group clauses too.

**Comment by parol-peter** [comment]
yeap. couse it checks presence of grouping by calling `group_values.any?` which for array of nils returns `false`

sometimes in more complicated stuffs value of grouping is being set dynamically eg. 
`g = how_to_group() # complicated method determining grouping strategy`
`i = what_to_include() # analogous method`
`relation = Model.includes(i).group(g) # model might be scoped/queried relation already`

i run onto such thing in gem wice_grid which takes g from options and default `g=nil`. Everything worked fine until I passed Model grouped already

so in such situation code for error throwing methods should look like
`relation = Model`
`relation = relation.group(g) if g_valid?(g)`
`relation = relation.includes(i) if i_valid?(i)`
to avoid nils (or other invalid values). this way whole pretty chaining idea goes to trash.

so IMHO chainable methods should use gracefull ignore of meaningless params rather than throwing errors that ruin whole chaining (might use logger.warn instead)

**Comment by rafaelfranca** [comment]
Still disagree. Nil is a invalid argument for these methods and ignoring it just may give people more confusion. And what if users were not expecting Nil and are getting those for some kind of wrong logic? 

The correct think to do is to check the argument before calling group. Silent ignoring is not an option.

**Comment by rafaelfranca** [comment]
I'll try to get more feedback. @sgrif @senny @matthewd

**Comment by meinac** [comment]
@rafaelfranca you are right, giving an argument error if the value is nil or blank will inform developers about their logic but we should change the behaviour of current group method it is currently ignores nil value while using it alone.

**Comment by senny** [comment]
As long as `group([])` works (it currently does) one could still chain by using `Array#compact`. I agree with @rafaelfranca that explicitly passing `nil` is invalid input and should not be silently ignored.

**Comment by parol-peter** [comment]
There are always two sides of the stuff. But according to my practise methods like this (chainable) deserve a bit different approach especially if meaningless value does not affect the very query

An option might be exposing some AR configuration to turn of error throwing in such case.
Another alternative might be group and group! like create and create!

RoR gives a pleasure of coding. It'd be nice to keep it this way even when you want to write something more complicated

The thing is not a default behaviour of AR but possibilities i've got to make it working my way

**Comment by senny** [comment]
Passing arrays should give you the chainability you are looking for. In the case of `nil` it's as likely that you have a programming mistake and it would be a pleasure (certainly for new programmers) if Rails told you so. For example `group(@psot)` instead of `group(@post)`.

**Comment by parol-peter** [comment]
yeap this is the one side of stuff which i understand but as I wrote previously:

`g = how_to_group() # complicated method determining grouping strategy might give nil`
`i = what_to_include() # analogous method`
`relation = Model.includes(i).group(g) # model might be scoped/queried relation already`

so in such situation code for error throwing methods should look like
`relation = Model`
`relation = relation.group(g) if g_valid?(g)`
`relation = relation.includes(i) if i_valid?(i)`
to avoid nils (or other invalid values). this way whole pretty chaining idea goes to trash.

**Comment by rafaelfranca** [comment]
You can make how_to_group always return valid input, so you still keep your
pretty chaining

**Comment by parol-peter** [comment]
but if no grouping is needed? this is determinied dynamically or by component configuration. it should return nil in that case.

i bet if you make query methods throw errors couple of gems will break up couse nils are pretty often when building queries programatically/from options on a fly

**Comment by senny** [comment]
@parol-peter should still work if `g` and `i` are either `[]` or `[:column]`. That can be easily achieved by using `Array().compact` as the return value of these methods.

**Comment by meinac** [comment]
@senny , @rafaelfranca what about giving blank string as an argument to group? `Model.group("")` I believe this should raise an error too.

**Comment by matthewd** [comment]
@meinac I think a blank string is Just Plain Wrong -- while it's friendlier to give a nice error message for a nil, which is a comparatively easy mistake to make, if you pass something totally unreasonable, it's fine to let it explode with whatever obscure error eventually turns up.

In this case, I'd argue that once you pass a string, you're accepting responsibility to ensure it contains a well-formed SQL fragment. There are a lot of ways you could get that wrong; a blank string is only one of them, and probably less likely than some others, at that.
